packet/flatcar-linux/kubernetes/calico/host-protection.yaml.tmpl

```diff
@@ -1,126 +1,109 @@
-# Customize inbound failsafe rules
-# https://docs.projectcalico.org/v3.5/getting-started/bare-metal/policy/failsafe
+# General
+# =======
+---
+# Disable default failsafe inbound rules
+apiVersion: crd.projectcalico.org/v1
+kind: FelixConfiguration
+metadata:
+  name: default
+spec:
+  failsafeInboundHostPorts: []
+
+# Host endpoints
+# ==============
+---
+${controller_host_endpoints}
+---
+${worker_host_endpoints}
+
+# Firewall policy
+# ===============
+---
+# SSH access
 apiVersion: crd.projectcalico.org/v1
 kind: GlobalNetworkPolicy
 metadata:
-  name: failsafe-ingress
+  name: ssh
 spec:
+  selector: has(host-endpoint)
   preDNAT: true
   applyOnForward: true
-  selector: has(host-endpoint)
   order: 0
   ingress:
-  # SSH access
   - action: Allow
     protocol: TCP
     source:
       nets: ${management_cidrs}
     destination:
       ports: [22]
-  # DHCP access
-  - action: Allow
-    protocol: UDP
-    source:
-      nets: ${node_public_ips}
-    destination:
-      ports: [68]
-  # BGP and etcd access
-  - action: Allow
-    protocol: TCP
-    source:
-      nets: ${node_public_ips}
-    destination:
-      ports:
-      # BGP access (Calico networking)
-      - 179
-      # etcd access
-      - 2379
-      - 2380
-      # etcd self-hosted service access
-      - 6666
-      - 6667
 ---
-# Allow traffic from IP addresses that are known to be cluster-internal including pods and services
+# kubectl access
 apiVersion: crd.projectcalico.org/v1
 kind: GlobalNetworkPolicy
 metadata:
-  name: allow-cluster-internal-ingress
+  name: kubectl
 spec:
-  order: 10
+  selector: nodetype == 'controller'
   preDNAT: true
   applyOnForward: true
+  order: 0
   ingress:
-    - action: Allow
-      source:
-        nets: ${cluster_internal_cidrs}
-  selector: has(host-endpoint)
+  - action: Allow
+    protocol: TCP
+    destination:
+      ports: [6443]
 ---
-# Deny traffic from any other sources
+# Traffic between cluster nodes
 apiVersion: crd.projectcalico.org/v1
 kind: GlobalNetworkPolicy
 metadata:
-  name: drop-other-ingress
+  name: inter-node-traffic
 spec:
-  order: 20
+  selector: has(host-endpoint)
   preDNAT: true
   applyOnForward: true
-  ingress:
-    - action: Deny
-  selector: has(host-endpoint)
----
-# Allow egress traffic through nodes' external interface
-apiVersion: crd.projectcalico.org/v1
-kind: GlobalNetworkPolicy
-metadata:
-  name: allow-outbound-external
-spec:
   order: 10
-  egress:
+  ingress:
     - action: Allow
-  selector: has(host-endpoint)
----
-# controller hosts endpoint
-${controller_host_endpoints}
----
-# worker hosts endpoint
-${worker_host_endpoints}
+      source:
```

Comment:
Actually, this refactor loosens the firewall rules between cluster nodes. I think we should keep them. We could split them into multiple policies though, to have one for Calico traffic, one for etcd etc. For example etcd traffic should be allowed only between controllers IIRC. Workers don't need to have access to them.

Reply:
>Actually, this refactor loosens the firewall rules between cluster nodes.

True, this slightly loosens the security there, to the same level of security you normally get on k8s cluster on AWS for example. It is very common practice to say "any worker can talk to any worker" and is, in my opinion, reasonable given the fact that these are all private IP addresses from a single CIDR inside a trusted data center with no other customers on the same hardware. On AWS the same is usually done by referencing a security group ID as a source. On Packet we don't have security groups and therefore the next best thing we can do is specify the project's CIDR. Ideas for improvement are welcome.

This should not create a security problem as long as we have a single cluster in a project. We can maybe say something about this in the docs.

>I think we should keep them.

We can't if we want to use auto scaling.

Reply:
Right. What about keeping it on CIDR level, but still have a list of ports/protocols then?

Reply:
My 2c: 
In AWS we don't have such open policy between the nodes: https://github.com/kinvolk/lokomotive-kubernetes/blob/master/aws/flatcar-linux/kubernetes/security.tf

If we are relaxing this bits compared not only to AWS but to our previous policy too, then we may want to have a good argument and document it (probably in the code too, as comment, to be very visible and no one accidentaly changes it in the future without understanding the current reasoning).

Reply:
>And what about keeping it on CIDR level, but still have a list of ports/protocols then, as I proposed?

@invidian you are very welcome to comment with a list of tested ports which we can whitelist. I would consider doing this in a separate PR as the focus of this PR is:

>...remove unnecessary rules and make the policy easier to understand and maintain.

Reply:
Why not just keep the list we had previously, which this PR removes? We can take it as a starting point and eventually improve later, if needed.

Reply:
@rata I am all for making security tighter. However, refining a list of ports and protocols to whitelist will take a while, and as the description of the PR states, the main goal here is to clean things up and not to change our level of enforcement. The latter would be better done in a dedicated PR.

As for loosening the security, in practice things stay largely the same (any node can talk to any node, same as before). The only additional implication of this change is that Packet hosts in the same project and the same facility can *also* talk to the k8s nodes. This will still be the case after refining a list of ports since we don't seem to have another way of expressing "just this cluster" in our policy in a maintainable way which allows auto scaling.

We may want to add to the docs that each cluster should reside in its own Packet project, without additional machines. Again, this is out of scope here IMO as I am not aiming to change our security policy, just clean up our currently-messy policy.

Reply:
>Why not just keep the list we had previously, which this PR removes? We can take it as a starting point and eventually improve later, if needed.

@invidian what list? We are currently not whitelisting ports between nodes:

https://github.com/kinvolk/lokomotive-kubernetes/blob/629e9bf884ae5f0be8cf60d75fa29fb38c57d974/packet/flatcar-linux/kubernetes/calico/calico-policy.yaml.tmpl#L42-L56

Reply:
@invidian @rata I've just taken another look at this rule. Actually we are not loosening security.

Current state:

https://github.com/kinvolk/lokomotive-kubernetes/blob/629e9bf884ae5f0be8cf60d75fa29fb38c57d974/packet/flatcar-linux/kubernetes/calico/calico-policy.yaml.tmpl#L42-L56

New state:

https://github.com/kinvolk/lokomotive-kubernetes/blob/f2b13530355d6a36d8333c0f1efcc32c1403aa28/packet/flatcar-linux/kubernetes/calico/host-protection.yaml.tmpl#L55-L69

Reply:
Oh, right, we had something whitelisted to go over public IPs and all that already was routed via private IPs was with this wide open rule. Sorry I missed it.

Then, it's fine by me as long as an issue to improve this and a policy equivalent to other providers is created :)

Reply:
Ah, I missed the fact that we were whitelisting over public IPs as well! All good then I guess :+1:

Reply:
Yeah, I confused the two as well. The public IP whitelisting is a leftover from before we switched to private IPs. In that case we had no choice but specify a list of IPs because Packet's public IPs aren't summarizable, i.e. they are from random pools.